src/Components/Components/src/NavigationManagerExtensions.cs

```diff
@@ -0,0 +1,938 @@
+            {
+                if (!pair.EncodedName.Span.Equals(encodedName, StringComparison.OrdinalIgnoreCase))
+                {
+                    newQueryStringBuilder.AppendParameter(pair.EncodedName.Span, pair.EncodedValue.Span);
+                }
+            }
+
+            return newQueryStringBuilder.UriWithQueryString;
+        }
+
+        /// <summary>
+        /// Returns a URI constructed from <see cref="NavigationManager.Uri"/> with multiple parameters
+        /// added, updated, or removed.
+        /// </summary>
+        /// <param name="navigationManager">The <see cref="NavigationManager"/>.</param>
+        /// <param name="parameters">The values to add, update, or remove.</param>
+        public static string UriWithQueryParameters(
+            this NavigationManager navigationManager,
+            IReadOnlyDictionary<string, object?> parameters)
+            => UriWithQueryParameters(navigationManager, navigationManager.Uri, parameters);
+
+        /// <summary>
+        /// Returns a URI constructed from <paramref name="uri"/> except with multiple parameters
+        /// added, updated, or removed.
+        /// </summary>
+        /// <param name="navigationManager">The <see cref="NavigationManager"/>.</param>
+        /// <param name="uri">The URI with the query to modify.</param>
+        /// <param name="parameters">The values to add, update, or remove.</param>
+        public static string UriWithQueryParameters(
+            this NavigationManager navigationManager,
+            string uri,
+            IReadOnlyDictionary<string, object?> parameters)
+        {
+            if (navigationManager is null)
+            {
+                throw new ArgumentNullException(nameof(navigationManager));
+            }
+
+            if (uri is null)
+            {
+                throw new ArgumentNullException(nameof(uri));
+            }
+
+            if (!TryRebuildExistingQueryFromUri(uri, out var existingQueryStringEnumerable, out var newQueryStringBuilder))
+            {
+                // There was no existing query, so there is no need to allocate a new dictionary to cache
+                // encoded parameter values and track which parameters have been added.
+                return UriWithAppendedQueryParameters(uri, parameters);
+            }
+
+            var parameterSourceCollection = CreateParameterSourceDictionary(parameters);
+
+            // Rebuild the query, updating or removing parameters.
+            foreach (var pair in existingQueryStringEnumerable)
+            {
+                if (parameterSourceCollection.TryGetValue(pair.EncodedName, out var source))
+                {
+                    if (source.AppendNextParameter(ref newQueryStringBuilder))
+                    {
+                        // We need to add the parameter source back into the dictionary since we're working on a copy.
+                        parameterSourceCollection[pair.EncodedName] = source;
```

Comment:
I'm probably misreading something here but how come we need to add it back? What code previously removes it?

Reply:
`source` is a `QueryParameterSource` here (which is a `struct`), so it's only a copy of the value in the dictionary:
```csharp
parameterSourceCollection.TryGetValue(pair.EncodedName, out var source)
```
`source.AppendNextParameter` mutates `source`, so the modified copy has to be added back into the dictionary. I'm definitely open to alternatives if you know of a better way to handle this.

Reply:
Oh, got it! Yes I see. A comment like `We have just mutated the struct value so we need to overwrite the copy in the dictionary` would be clearer to me at least. I thought that *add it back* meant *it has been removed* not *overwrite it*.

Reply:
Ah, I can definitely see how that could be confusing now. I'll improve that comment too. Thanks!